### ISSUE TITLE
Implement RFC0052: Graceful Stack Upgrades

### ISSUE THREAD

**Issue by ryanmoran** [paketo-buildpacks/ruby#819]
[RFC](https://github.com/paketo-buildpacks/rfcs/blob/main/text/0052-graceful-stack-upgrades.md)

## Summary

All Paketo Buildpacks should support graceful stack upgrades between the current Bionic and Jammy stacks. For most buildpacks, this should not require any changes, but for some buildpacks, layers created in previous builds that would normally be reused should be recreated if the stack ID changes.

## Action Item
Consider what types of buildpacks would be impacted by a stack upgrade and make any relevant changes that would help to mitigate issues caused by that upgrade. If there is no work to be done, please comment on this issue explaining the context that helped to arrive at that conclusion.

**Comment by fg-j** [comment]
Blocked by https://github.com/paketo-buildpacks/ruby/issues/793

**Comment by sophiewigmore** [comment]
Buildpacks of interest:
- Bundle Install
- Rails Assets

**Comment by sophiewigmore** [comment]
Investigation so far has lead me to believe we will need cache invalidation for the Bundle Install buildpack (and likely the Rails Assets buildpack, but more investigation on that to come)
Steps:
1. Build https://github.com/paketo-buildpacks/rails-assets/tree/main/integration/testdata/6.1 on Bionic
2. Runs successfully
3. Re-build successfully the same app on Jammy, Bundle Install buildpack gems are reused, Rails Assets buildpack assets are reused
4. App fails to run:
```
bundler: failed to load command: puma (/layers/paketo-buildpacks_bundle-install/launch-gems/ruby/3.1.0/bin/puma)
/layers/paketo-buildpacks_bundle-install/launch-gems/ruby/3.1.0/gems/puma-5.6.5/lib/puma.rb:13:in `require': libssl.so.1.1: cannot open shared object file: No such file or directory - /layers/paketo-buildpacks_bundle-install/launch-gems/ruby/3.1.0/gems/puma-5.6.5/lib/puma/puma_http11.so (LoadError)
```

**Comment by sophiewigmore** [comment]
Bundle install gem cache reuse **works** when building basic apps like our [Unicorn sample](https://github.com/paketo-buildpacks/samples/tree/main/ruby/unicorn) and the [Passenger sample](https://github.com/paketo-buildpacks/samples/tree/main/ruby/passenger).

**Comment by sophiewigmore** [comment]
Modifying the Bundle Install buildpack to clear the cached layer on stack upgrade, I haven't observed any other failures upgrading related to the Rails Assets buildpack. I believe only Bundle Install will need to be updated.

**Comment by sophiewigmore** [comment]
Final update - it turns out graceful stack upgrades only fail (in the cases I investigated) when doing the original build on the Bionic Full stack, and then upgrading to Jammy. This is because of the version of `openssl` in the Bionic Full stack. Upgrades succeeded across the board in my exploration when starting with the Bionic Base stack. AS a result, the bundle install cache will need to be cleared on all stack upgrades in order to be safe

**Comment by sophiewigmore** [comment]
Changes to bundle install released in Bundle Install buildpack v0.6.0 and Ruby language family buildpack v0.21.0
